Add JSON Schema media types (#19)
ietf-wg-httpapi/mediatypes: draft-ietf-httpapi-rest-api-mediatypes.md

Change controller:  n/a

## JSON Schema Media Types

JSON Schema is a declarative domain-specific language for validating and
annotating JSON documents (see {{jsonschema}}).

This document registers the media types associated with JSON Schema.

There are many dialects of JSON Schema in wide use today. The JSON Schema
maintainers have released several dialects including draft-04, draft-07, and

> draft 2020-12. There are also several third-party JSON Schema dialects in wide

Reviewer: no 2019-09?

Reviewer 2: It's arguaeable that the three listed are the most prominant and well used.

Author: It wasn't intended to be an exhaustive list. I didn't include draft-06 either. I mentioned the most widely adopted versions plus 2020-12 which we expect is destined to be widely adopted because of OpenAPI 3.1. Should I include more? Should I be more clear that the list is non-exhaustive?

Reviewer 3: +1 for non-exhaustive list.

use including the ones defined for use in OpenAPI and MongoDB.

This specification defines little more than how to identify the dialect while
leaving most of the semantics of the schema up to the dialect to define. Clients
MUST use the following order of precedence for determining the dialect of a
schema.

* The `$schema` keyword ({{schema-keyword}})
* The "schema" media type parameter ({{schema-parameter}})
* The context of the enclosing document. This applies only when a schema is
  embedded within a document. The enclosing document could be another schema in
  the case of a bundled schema or it could be another type of document that
  includes schemas such as an OpenAPI document.
* If none of the above result in identifying the dialect, client behavior is

>   undefined.

Reviewer 2: I've been wondering if ideally we want undeclared dialects to result in throwing an error.
SO many times people run into problems where they are using a library which doesn't support the dialect they are trying to use.
If we can't say MUST, I would hope we could agree on RECOMMENDED.

Author: I would love it to be required, but the spec has never required it and existing implementations have different behaviors, so I think we need to allow anything.

As for making a recommendation, I generally don't think a specification is the right place to be suggesting a design decision, but if people think it's important enough to say something, I'm ok with it. Personally, I think that saying something is "undefined" in a specification is very strong statement that people should not be relying on that behavior. Users should avoid it at all costs and implementers should warn or error in those cases to let the user know they are in dangerous territory.

Reviewer 3: If this is already in 2020-12 specification, since it's not strictly related to the registration fields, I think we could just reference the relevant section of the jsonschema specifications.

> I generally don't think a specification is the right place to be suggesting a design decision,

imho that depends if those decisions have impact on interoperability and security. In this case, while I am not sure the media type registration is the right place for this information - but we could ask to the WG - probably it's worth addressing the issue.

Let's move this specific issue to another PR so that we can move on!

Reviewer 2: > I would love it to be required, but the spec has never required it and existing implementations have different behaviors, so I think we need to allow anything. - @jdesrosiers 

Agreed. It was more a comment for "in future versions of the sepc I would like to...". But we have to deal with the here and now.l

### The "$schema" Keyword {#schema-keyword}

The `$schema` keyword is used as a JSON Schema dialect identifier. The value of
this keyword MUST be a URI {{!RFC3986}}. This URI SHOULD identify a meta-schema
that can be used to validate that the schema is syntactically correct according
to the dialect the URI identifies.

The dialect SHOULD define where the `$schema` keyword is allowed and/or
recognized in a schema, but it is RECOMMENDED that dialects do not allow the
schema to change within the same Schema Resource.

### Identifying a Schema via a Media Type Parameter {#schema-parameter}

> Media types MAY allow for a `schema` media type parameter, which gives HTTP
> servers the ability to perform Content-Type Negotiation based on schema
> identifier. The media type parameter MUST be a URI-reference {{!RFC3986}}.

Reviewer 3: ```suggestion
Media types MAY allow for a `schema` media type parameter
to support content negotiation (see Section 12 of {{SEMANTICS}})
. The media type parameter MUST be a URI-reference {{!RFC3986}}.
```

Author: :+1: for including a reference, but we can't reference an I-D. I'll see if I can come up with the correct reference.

The `schema` parameter identifies a schema that provides semantic information
about the resource the media type represents. When using the
`application/schema+json` media type, the `schema` parameter identifies the
dialect of the schema the media type represents.

The `schema` URI is opaque and SHOULD NOT automatically be dereferenced. Since
`schema` doesn't necessarily point to a network location, the "describedby"
relation is used for linking to a downloadable schema.

> In HTTP, the media-type parameter would be sent inside the Content-Type header:
> 
> ~~~
> Content-Type: application/schema-instance+json; schema="/schemas/dog"
> ~~~
> 
> ~~~
> Content-Type: application/schema+json; schema="https://json-schema.org/draft/2020-12/schema"
> ~~~
> 
> Media type parameters are also used in HTTP's Accept request header:
> 
> ~~~
> Accept: application/schema-instance+json; schema="/schemas/dog",
>         application/schema-instance+json; schema="/schemas/cat"
> ~~~
> 
> ~~~
> Accept: application/schema+json; schema="https://json-schema.org/draft/2020-12/schema",
>         application/schema+json; schema="http://json-schema.org/draft-07/schema#"
> ~~~

Reviewer 3: ```suggestion
An example of content negotiation where a user agent requests
two possible schema instances, and the server replies with
the one with a specific schema.

Request:

~~~ http-message
GET /schema HTTP/1.1
Host: example
Accept: application/schema-instance+json; schema="/schemas/dog",
             application/schema-instance+json; schema="/schemas/cat"
~~~

Response:

~~~ http-message
HTTP/1.1 200 Ok
Content-Type: application/schema-instance+json; schema="/schemas/dog"

{ ... }
~~~

In the following example, the user agent is willing to accept two possible
version of json-schema, and the server replies with the latest one.

Request:

~~~ http-message
GET /schema HTTP/1.1
Host: example
Accept: application/schema+json; schema="https://json-schema.org/draft/2020-12/schema",
        application/schema+json; schema="http://json-schema.org/draft-07/schema#"
~~~

Response:

~~~ http-message
HTTP/1.1 200 Ok
Content-Type: application/schema+json; schema="https://json-schema.org/draft/2020-12/schema"

{ ... }
~~~

```

Author: This section definitely needs improvement. I was going to come back to this next time I had time for a revision. However, this does not correctly describe how the schema instance media type works. I'll work on it.

### Linking to a Schema {#schema-linking}

It is RECOMMENDED that instances described by a schema provide a link to a
downloadable JSON Schema using the link relation `describedby`, as defined by
Linked Data Protocol 1.0, section 8.1 {{!W3C.REC-ldp-20150226}}.

> In HTTP, such links can be attached to any response using the Link header
> {{!RFC8288}}. An example of such a header would be:

Reviewer 3: ```suggestion
In HTTP, such links can be attached to any response using the `Link` header field
{{!LINK=RFC8288}}. An example of `Link` would be:
```

~~~
Link: <https://example.com/my-hyper-schema#>; rel="describedby"
~~~

### Fragment Identifiers {#schema-fragment}

Two fragment identifier structures are supported: JSON Pointers and plain-names.

The use of JSON Pointers as URI fragment identifiers is described in RFC 6901
{{!RFC6901}}. Fragment identifiers that are empty or start with a `/`, MUST be
interpreted as JSON Pointer fragment identifiers.

Plain-name fragment identifiers reference locally named locations in the
document. The dialect determines how plain-name identifiers map to locations
within the document. All fragment identifiers that do not match the JSON Pointer
syntax MUST be interpreted as plain name fragment identifiers.

### Media Type application/schema+json {#schema-json}

The `application/schema+json` media type represents JSON Schema. This schema can
be an official dialect or a third-party dialect. The following information
serves as the registration form for the `application/schema+json` media type.

**Type name**: application

**Subtype name**: schema+json

**Required parameters**: N/A

**Optional parameters**:

* **schema**: A URI identifying the JSON Schema dialect the schema was written

>   for. If this value conflicts with the value of the `$schema` keyword in the

Reviewer 3: This text is more clear than the normative one above! :)

  schema, the `$schema` keyword takes precedence.

>   Unrecognized parameters SHOULD be ignored.

Reviewer 3: not sure we can use normative language in the media type registration. Instead, reference the normative part in this document.

Author: I'm not sure why normative language would be inappropriate here, but I'm fine with changing it. This probably isn't something that we even need to say explicitly. Maybe we just remove the line.

Reviewer 4: If anything it should (ha) be a MUST

Reviewer 3: It seems @richsalz confirms normative language is ok. 
Since it requires mentioning bcp14+ which is not mentioned around I thought it wasn't.

I agree with @richsalz that's a MUST.

responses, IDEs that valid configuration files, databases that store JSON, and
more.

**Fragment identifier considerations**: See {{schema-fragment}}

**Additional information**:

  * **Deprecated alias names for this type**: N/A

  * **Magic number(s)**: N/A

  * **File extension(s)**: json, schema.json

  * **Macintosh file type code(s)**: N/A

**Person and email address to contact for further information**: See Authors'
Addresses section.

**Intended usage**: COMMON

**Restrictions on usage**: N/A.

**Author**: See Authors' Addresses section.

**Change controller**: N/A

### Media Type application/schema-instance+json {#schema-instance-json}

The `application/schema-instance+json` media type is an extension of the
{{JSON}} media type that just adds the `schema` media type parameter and
fragment identification. The following information serves as the registration
form for the `application/schema-instance+json` media type.

**Type name**: application

**Subtype name**: schema-instance+json

**Required parameters**: N/A

**Optional parameters**:

* **schema**: A URI identifying a JSON Schema that provides semantic information
  about this JSON representation.

>   Unrecognized parameters SHOULD be ignored

Reviewer 3: Taken from [json](https://www.iana.org/assignments/media-types/application/json)

```suggestion
  Unrecognized parameters have no effect on compliant recipients.
```
